### PR TITLE
Fix join spill race condition

### PR DESCRIPTION
**Pull request by sopel39**
SpilledLookupSourceHandle needs to be disposed
immediately when join operators finished
processing probe data. Otherwise SpilledLookupSourceHandle
is never disposed as it's not registered in PartitionedConsumption.
This prevents HashBuilderOperator from finishing.

Fixes: https://github.com/trinodb/trino/issues/7454

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java [hunk 1]
```diff
@@ -212,7 +212,10 @@ public void setPartitionSpilledLookupSourceHandle(int partitionIndex, SpilledLoo
 
         lock.writeLock().lock();
         try {
-            if (destroyed.isDone()) {
+            if (partitionedConsumption.isDone() || destroyed.isDone()) {
+                // Either:
+                // 1. join operators already finished consuming probe data and are unspilling
```
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java:217]: perhaps the condition should be on `partitionsNoLongerNeeded.isDone()`
then you don't need `|| destroyed.isDone()` part anymore.

**Comment by sopel39** [comment]
ac @findepi

File: core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java]: > partitionsNoLongerNeeded is not setup in finishProbeOperator

it seems to be set here https://github.com/trinodb/trino/blob/03d8e5abc686c5c4a9f96ca14db07e2aed880174/core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java#L326

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java]: > nope. They could have already processed all probe data because at that time HashBuilder was not spilled. It's not short-circuit.

if `setPartitionSpilledLookupSourceHandle` is being called, it seems to mean the `lookupSourceSupplier` was not set yet and than means Probe should be waiting here
on `lookupSourceProviderFuture.isDone()` being `false` in 
https://github.com/trinodb/trino/blob/4421ee68994e6c524882043b104ccb4d53e9312e/core/trino-main/src/main/java/io/trino/operator/LookupJoinOperator.java#L124

what am i missing?

[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java]: 1. `lookupSourceProviderFuture` can unblock on non-spilled lookup source
2. This method is about `io.trino.operator.SpilledLookupSourceHandle`. I believe a new lookup source is created when spill happens.

File: core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java [hunk 2]
```diff
@@ -74,6 +74,9 @@
     @GuardedBy("lock")
     private final SettableFuture<?> destroyed = SettableFuture.create();
 
+    @GuardedBy("lock")
+    private boolean probeFinished;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java:78]: per https://github.com/trinodb/trino/pull/7455#discussion_r605756896 can you help me understand why do we need a new field?

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSourceFactory.java:78]: `io.trino.operator.PartitionedLookupSourceFactory#finishProbeOperator` can exist in two places. Both need to be protected from race.

**Comment by findepi** [comment]
(discussed offline)

File: core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java [hunk 3]
```diff
@@ -2110,7 +2110,9 @@ private PhysicalOperation createLookupJoin(
             PhysicalOperation probeSource = probeNode.accept(this, context);
 
             // Plan build
-            boolean spillEnabled = isSpillEnabled(session) && node.isSpillable().orElseThrow(() -> new IllegalArgumentException("spillable not yet set"));
+            boolean spillEnabled = isSpillEnabled(session)
+                    && node.isSpillable().orElseThrow(() -> new IllegalArgumentException("spillable not yet set"))
+                    && probeSource.getPipelineExecutionStrategy() == UNGROUPED_EXECUTION;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java:2115]: Would it make sense to extract this to a separate commit?

File: core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java [hunk 4]
```diff
@@ -2323,8 +2325,11 @@ private OperatorFactory createLookupJoin(
             List<Integer> probeJoinChannels = ImmutableList.copyOf(getChannelsForSymbols(probeSymbols, probeSource.getLayout()));
             OptionalInt probeHashChannel = probeHashSymbol.map(channelGetter(probeSource))
                     .map(OptionalInt::of).orElse(OptionalInt.empty());
-            OptionalInt totalOperatorsCount = context.getDriverInstanceCount();
-            checkState(!spillEnabled || totalOperatorsCount.isPresent(), "A fixed distribution is required for JOIN when spilling is enabled");
+            OptionalInt totalOperatorsCount = OptionalInt.empty();
+            if (spillEnabled) {
+                totalOperatorsCount = context.getDriverInstanceCount();
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java:2330]: nit: extract?

**Review by findepi — APPROVED** [review]
